opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -208,8 +209,38 @@ def set_attribute(self, key: str, value: types.AttributeValue) -> None:
         if has_ended:
             logger.warning("Setting attribute on ended span.")
             return
+
+        if not isinstance(value, (int, float, bool, str, list, tuple)):
+            logger.warning("invalid type for attribute value")
+            return
+        if isinstance(value, (list, tuple)) and len(value) > 0:
+            return_code = self._check_sequence(value)
+            if return_code is not None:
+                logger.warning("%s in attribute value sequence", return_code)
+                return
+
         self.attributes[key] = value
```

Comment:
There's a potential edge case where AttributeValues that are lists can be mutated afterward, resulting in invalid types that exporters will run into. 

I've added a followup ticket on that here: https://github.com/open-telemetry/opentelemetry-python/issues/352

Reply:
That's a great catch. Would adding a copy of the list rather than the original list resolve this?

Reply:
We could store the copies of the values in tuples. Now that you mention this, instead of accepting lists or tuples we should accept [sequences](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence).

Reply:
yeah, I think that's a good solution. Would be good to make a separate PR for that.